### DISCUSSION TITLE
Tuple form of input x?

### DISCUSSION THREAD

**Discussion by Nokimann** [pyg-team/pytorch_geometric#3552] [Q&A] [answered]
Modules in `torch_geometric.nn.conv` can get the tuple form of input `x`.
Each means x_src and x_tgt (?)

1. Can we use them, with different dimensions of x_src and x_tgt, if we just set the tuple of `in_channels`?

2. If the above true, does `[j, i]` pair in `edge_index` means "from `j-th row of x_src` to `i-th row of x_tgt`"?

**Comment by rusty1s** [comment, marked as answer]
1. Yes, you can see [here](https://pytorch-geometric.readthedocs.io/en/latest/notes/cheatsheet.html) which operators support "bipartite" message passing flow.
2. Yes :)

**Reply by Nokimann** [reply]
Thank you as always.

**Reply by ZIKEYUAN** [reply]
I have a question about input as tuple. I would like to ask what is the difference between input x as tuple and input as tensor? I found that some GNN models do not need to convert x to tuple but some do. I am very confused about this.

**Reply by rusty1s** [reply]
A tuple of node features, e.g., `(x_src, x_dist)` represents a bipartite message passing flow where source node features `x_src` are passed to destination nodes and aggregated to `x_dst` to compute a new destination node representation. If it is passed as a tensor, we assume that both `x_src` and `x_dst` are equal.

**Reply by ZIKEYUAN** [reply]
Thank you very much!
